Fate specification (#279)
aeternity/protocol: contracts/fate.md

## Types

The machine supports the following types:
* Integer (signed arbitrary precision integers) [I]
* Boolean (true | false) [B]
* Addresses (A pointer into the state tree) [A]
* Strings (utf8 encoded byte arrays) [S]
* Tuples (), ('a), ('a, 'b, ...) [T]
* Lists: Cons cells (‘a) | Nil (Ʇ)  [L],[C],[Ʇ]
* Maps (‘a, ‘b),  (Key: ‘a -> Value: ‘b)  ‘a not a map [M]
* Variant Types, (| Size | Tag | Elements  |)  [V]
* Bits <Bits> or !<Bits>
* TypeRep [R]

These types can be used as argument and return types of functions and
also as the type of the elements in the contract store.

### Integer

The type Integer, is used for any integer, infinitely large or small
(below zero).

Examples:

```
 0
 1
-1
589465789334
-51315353513513131656462262
```

### Boolean

The Boolean type only has two values, true or false.

Examples (all of them):
```
 true
 false
```

Operations on booleans include:
*   and, or, not
*   conditional jumps

### Addresses

Values of the type Address are 32-bytes (256-bits) pointers to
entities on the chain (Accounts, Oracles, Contracts, etc)

Examples:
```
<<ak_2HNb8bivUoJTcaxD6VKDy2wPHvTuQgQRJ3dkF9RSyVFqBkMneC>>
```

### Strings

> Strings are byte arrays od UTF-8 encoded unicode characters. (These

Reviewer: ```suggestion
Strings are byte arrays of UTF-8 encoded unicode characters. (These
```

> Strings are byte arrays od UTF-8 encoded unicode characters. (These

Reviewer: Re UTF-8: is it enforced in the language or in the VM?

I expect that at the lowest level the string is represented as a sequence of bytes. But not all sequence of bytes are valid UTF-8:
```
$ erl
Erlang/OTP 20 [erts-9.3.3.9] [source] [64-bit] ...
$ erl
Erlang/OTP 20 [erts-9.3.3.9] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe] [kernel-poll:false]
Eshell V9.3.3.9  (abort with ^G)
1> lists:map(fun unicode:characters_to_list/1, [<<2#00100100>>, <<2#11000010, 2#10100010>>]).
["$","¢"]
2> lists:map(fun unicode:characters_to_list/1, [<<2#10000000>>, <<2#10000010, 2#10100010>>]).
[{error,[],<<128>>},{error,[],<<130,162>>}]
```

Does it mean that all points where strings are instantiated in the VM are checked to be valid UTF-8? Or rather an exception is raised by the VM when the assumption of UTF-8 validity is enforced?

Also, I need to understand better if the validity of a string according to UTF-8 changes depending on the version of Unicode.

can be used as non-indexable arrays of bytes.)

Examples:
```
“Hello world”
  “eof”
```

### Tuples

Tuples have an arity indicating the number of elements in the
tuple. Each element in the tuple has a type. The 0-tuple is also

> sometimes called unit.

Reviewer: ```suggestion
sometimes called unit.
The maximum number of elements in the tuple is 255 (included). 
```

Examples:
```
 {}                Type: unit
 {1, 2}            Type: Tuple(Integer, Integer)
 {“foo”, 42, true} Type: Tuple(String, Integer, Boolean)
```

### Lists

Lists are monomorphic series of values. Hence lists have a type. There
is also an empty list (or Nil).

Examples:
```
[]                   Type: Nil
[1, 2, 3]            Type: List(Integer)
[true, true, false]  Type: List(Boolean)
```

### Maps

Maps are monomorphic key value stores, that is the key is always the
same type for all elements in a map and the value has the same type in

> all mappings in a map.  The key kan have any type except a map.  The

Reviewer: ```suggestion
all mappings in a map.  The key can have any type except a map.  The
```

> value kan have any type including a map.

Reviewer: ```suggestion
value can have any type including a map.
```

Examples:
```
#{ (1, “foo”), (2, “bar”) }
```
Type: Map(Integer, String)
```
#{ (“Fruit prices”, #{ (“banana”, 42), (“apple”, 12)}),
   (“Stock prices”, #{(“Apple”, 142), (“Orange”, 112)}) }
```

### Variant Types

The variant type is a type consisting of a size and a tag (index)
where each tag represents a series of values of specified types. For
example you could have the Option type which could be None or

> Some(Integer). The size of the variant is 2 (there are two varaints),

Reviewer: ```suggestion
Some(Integer). The size of the variant is 2 (there are two variants),
```

The value None would be indicated by tag 0. The value
Some(42) would be represented by tag 1 followed by the integer.

Examples:
```
  (| 2 | 0 | [] |)      ;; None
  (| 2 | 1 | (42) |)    ;; Some(42)

  (| 4 | 3 | (42, “foo”, true) |) ;; Three(42, "foo", true)
```

> Note that the tuple syntax for the elements does not indicate a Fate

Reviewer: Square brackets `[]` here, parentheses `(` below. May surprise reader.

are stored in the byte following the opcode.
Operand specifiers for operations with 5 to 8 arguments
are stored in the two bytes following the opcode.
Note that for many operation the first argument (arg0) is
the destination for the operation.

| value: | arg3 | arg3 | arg2 | arg2 | arg1 | arg1 | arg0 | arg0 |
| ---    | ---  | ---  | ---  | ---  | ---  | ---  | ---  | ---  |
| bit:   |    7 |   6  |    5 |    4 |    3 |    2 |    1 |   0  |

| value: | arg8 | arg7 | arg6 | arg6 | arg5 | arg5 | arg4 | arg4 |
| ---    | ---  | ---  | ---  | ---  | ---  | ---  | ---  | ---  |
| bit:   |    7 |   6  |    5 |    4 |    3 |    2 |    1 |   0  |

Unused arguments should be set to 0.

E.g. an a :=  immediate + a would have the bit
pattern 00 11 00 00.

### Operations

| OpCode | Name | Args | Description |
| ---    | ---  | ---  |        ---  |
| 0x0 | 'RETURN' |  | Return from function call pop stack to arg0. The type of the retun value has to match the return type of the function. |
| 0x1 | 'RETURNR' | Arg0 | Return from function call copy Arg0 to arg0. The type of the retun value has to match the return type of the function. |

> | 0x2 | 'CALL' | Identifier | Call given function with args on stack. The types of the arguments has to match the argument typs of the function. |

Reviewer: ```suggestion
| 0x2 | 'CALL' | Identifier | Call given function with args on stack. The types of the arguments has to match the argument types of the function. |
```

> | 0x3 | 'CALL_R' | Arg0 Identifier | Remote call to given contract and function.  The types of the arguments has to match the argument typs of the function. |

Reviewer: ```suggestion
| 0x3 | 'CALL_R' | Arg0 Identifier | Remote call to given contract and function.  The types of the arguments has to match the argument types of the function. |
```

> | 0x4 | 'CALL_T' | Identifier | Tail call to given function. The types of the arguments has to match the argument typs of the function. And the return type of the called function has to match the type of the current function. |

Reviewer: ```suggestion
| 0x4 | 'CALL_T' | Identifier | Tail call to given function. The types of the arguments has to match the argument types of the function. And the return type of the called function has to match the type of the current function. |
```

> | 0x5 | 'CALL_TR' | Arg0 Identifier | Remote tail call to given contract and function. The types of the arguments has to match the argument typs of the function. And the return type of the called function has to match the type of the current function. |

Reviewer: ```suggestion
| 0x5 | 'CALL_TR' | Arg0 Identifier | Remote tail call to given contract and function. The types of the arguments has to match the argument types of the function. And the return type of the called function has to match the type of the current function. |
```

| 0x46 | 'VARIANT_ELEMENT' | Arg0 Arg1 Arg2 | Arg0 := element number Arg2 from variant Arg1. |
| 0x47 | 'BITS_NONEA' |  | accumulator := empty bitmap. |
| 0x48 | 'BITS_NONE' | Arg0 | Arg0 := empty bitmap. |
| 0x49 | 'BITS_ALLA' |  | accumulator := full bitmap. |
| 0x50 | 'BITS_ALL' | Arg0 | Arg0 := full bitmap. |
| 0x51 | 'BITS_ALL_N' | Arg0 Arg1 | Arg0 := bitmap with Arg1 bits set. |
| 0x52 | 'BITS_SET' | Arg0 Arg1 Arg2 | Arg0 := set bit Arg2 of bitmap Arg1. |
| 0x53 | 'BITS_CLEAR' | Arg0 Arg1 Arg2 | Arg0 := clear bit Arg2 of bitmap Arg1. |
| 0x54 | 'BITS_TEST' | Arg0 Arg1 Arg2 | Arg0 := true if bit Arg2 of bitmap Arg1 is set. |
| 0x55 | 'BITS_SUM' | Arg0 Arg1 | Arg0 := sum of set bits in bitmap Arg1. Exception if infinit bitmap. |
| 0x56 | 'BITS_OR' | Arg0 Arg1 Arg2 | Arg0 := Arg1 v Arg2. |
| 0x57 | 'BITS_AND' | Arg0 Arg1 Arg2 | Arg0 := Arg1 ^ Arg2. |
| 0x58 | 'BITS_DIFF' | Arg0 Arg1 Arg2 | Arg0 := Arg1 - Arg2. |
| 0x59 | 'ADDRESS' | Arg0 | Arg0 := The current contract address. |
| 0x5a | 'BALANCE' | Arg0 | Arg0 := The current contract address. |
| 0x5b | 'ORIGIN' | Arg0 | Arg0 := Address of contract called by the call transaction. |

> | 0x5c | 'CALLER' | Arg0 | Arg0 := The address that signed the call transaction. |

Reviewer: As discussed with @happi and @gorillainduction , may need review.

Reviewer 2: One minor thing, but with generalised accounts, the »signed« could be no longer accurate, so it might be good to start thinking of a more general notion of »authorised«.

| 0x6b | 'DEACTIVATE' |  | Mark the current contract for deactication. |
| 0x6c | 'SPEND' | Arg0 Arg1 | Transfer Arg0 tokens to account Arg1. (If the contract account has at least that many tokens. |
| 0x6d | 'ORACLE_REGISTER' | Arg0 Arg1 Arg2 Arg3 Arg4 Arg5 | Mark the current contract for deactication. |
| 0x6e | 'ORACLE_QUERY' |  |  |
| 0x6f | 'ORACLE_RESPOND' |  |  |
| 0x70 | 'ORACLE_EXTEND' |  |  |
| 0x71 | 'ORACLE_GET_ANSWER' |  |  |
| 0x72 | 'ORACLE_GET_QUESTION' |  |  |
| 0x73 | 'ORACLE_QUERY_FEE' |  |  |
| 0x74 | 'AENS_RESOLVE' |  |  |
| 0x75 | 'AENS_PRECLAIM' |  |  |
| 0x76 | 'AENS_CLAIM' |  |  |
| 0x77 | 'AENS_UPDATE' |  |  |
| 0x78 | 'AENS_TRANSFER' |  |  |
| 0x79 | 'AENS_REVOKE' |  |  |
| 0x7a | 'ECVERIFY' |  |  |
| 0x7b | 'SHA3' |  |  |
| 0x7c | 'SHA256' |  |  |
| 0x7d | 'BLAKE2B' |  |  |
| 0xf9 | 'DUMMY7ARG' | Arg0 Arg1 Arg2 Arg3 Arg4 Arg5 Arg6 | Temporary dummy instruction to test 7 args. |
| 0xfa | 'DUMMY8ARG' | Arg0 Arg1 Arg2 Arg3 Arg4 Arg5 Arg6 Arg7 | Temporary dummy instruction to test 8 args. |
| 0xfb | 'ABORT' | Arg0 | Abort execution (dont use all gas) with error message in Arg0. |

> | 0xfc | 'EXIT' | Arg0 | Abort execution (use upp all gas) with error message in Arg0. |

Reviewer: ```suggestion
| 0xfc | 'EXIT' | Arg0 | Abort execution (use up all gas) with error message in Arg0. |
```

* Trace / debug
* Side effects (Tx : contract)
* Non spendables (fun and contract)
* Events in Sophia (Roma)    (Hans)
* Test compatible versions
* Stateful, Pure recognized by Sophia compiler
* Recursive types in Sophia
* Sophia inheritance (compile time code organization)
* Polymorphism
* Standard Lib
* FATE spec (Erik)
** ABI
** Instruction set
** State
* Primops spec  (Tobias)
* VM
* Sophia backend -> FATE  (Ulf)
* Refactor Tx -> Primop
* Gas pricing
* State tree (cache / handling) + side effect checking
* Implement instruction set
* De-serialization
* New instruction hash (Blake)
* Hash on arbitrary VM objects
** String hash
** Phash

Done:

* Sophia stand alone (Robert)
* AEVM opcodes  -> repo  (Robert)
* AEVM /Sophia ABI code in separate library/cleanup
* Sophia server http interface (separate repo)
* Sophia contract interface (also called “ABI”) 45 min
* Hard fork support
* VM version handling

# Appendix 1: Fate instruction set and serialization

Most instructions are encoded as one byte opcode (7-bits), one byte
operand specifiers and 0 to 3 variable size arguments.

The 7th bit (counting from 0) is used for future extensions beyond 127 instructions.

> Encoding of types

Reviewer: Markup in this section beyond this point would benefit from review.

Immediates in code (constants) are encoded by the instruction type and then depending on the type of the immediate as follows:
Integer:  Encoded with RLP where the first type bit is reused as a sign bit.
Boolean: One byte, false: 0, true: 1.
Addresses: 32 bytes
Byte arrays: RLP encoded.

> Tuples: 1 byte length, [encoded types], [encoded elements]

Reviewer: I see 1 byte length for tuple.

integer" below we mean an integer encoded as the smalles big endian byte array
then this byte array is RLP encoded.

```
integer      : sxxxxxx0 : 6 bit integer with sign bit when integer < 64
string       : xxxxxx01 | size | [bytes] : when 0 < size < 64
string       : 00000001 | RLP encoded byte array : when size >= 64
list         : 00000011 | RLP encoded (length - 16) | [encoded elements] : when length >= 16
list         : xxxx0011 | [encoded elements] : when 0 < length < 16, xxxx = length
             : xxxx0111 : FREE (For typedefs in future)
tuple        : 00001011 | RLP encoded (size - 16) | [encoded elements] : when size >= 16
tuple        : xxxx1011 | [encoded elements] : when 0 < size(tuple) < 16
map          : 00101111 | RLP encoded size | [encoded key, encoded value]
tuple        : 00111111 : when size(tuple) == 0
bits         : 01001111 | RLP encoded integer when bits are finate
string       : 01011111 : when size == 0
integer      : 01101111 | RLP encoded (integer - 64) when size >= 64 and integer >= 0
false        : 01111111
             : 10001111 : FREE (Possibly for bytecode in the future.)
address      : 10011111 | [32 bytes]

> varaint      : 10101111 | RLP encoded variant size field | encoded tag | encoded values

Reviewer: ```suggestion
variant      : 10101111 | RLP encoded variant size field | encoded tag | encoded values
```
?

list         : 10111111 : when length == 0
bits         : 11001111 | RLP encoded integer : when bits are in infinite
map          : 11011111 : when size == 0
integer      : 11101111 | RLP encoded ((- integer) - 64)
true         : 11111111
```

# Notation

Some notes on notation and definition of terms used in this document.

## Bits

Bits are counted from the least significant bit, starting on 0. When
bits are written out they are written from the most significant bit to
the least significant bit.

Example the number 1 in a byte would be written in binary as:
```

> 0000001

Reviewer: ```suggestion
00000001
```
?

Unless we define byte as 7 bits :)

records and store each field under a separate key.

#### The state tree

Contains all accounts, oracles and contracts on the chain. Most values
such as account balances can be read directly through specific load
instructions. Some values can be changed through transaction
instructions.

#### The local storage

The local storage is a stack of key value storages. That is, a list of
maps. When reading a key, the machine first looks in the first map of
the list. If a key is found and the type is correct the value is
returned. If the type is incorrect an exception is thrown. If no value
is found in the first map, the machine will look in the next map in
the list and so on. If the key does not exist in any of the maps an
exception is thrown.

The key is a an integer (32 bit).

#### The accumulator stack

The accumulator stack is a stack of typed values. The top of the stack
can be used as an argument to an operation. Values can be pushed to
the stack and popped from the stack.

#### The execution stack

The execution stack contain return addresses as a tuple in the form of
{contract address, function hash, and basic block index}.

#### The event stream

> A contract can emit events similar to the EVM.

Reviewer 3: This is rather cryptic... not sure people are familiar with EVM. Potentially add a line on the FATE primitive that is used to do so?

```

### Strings

Strings are byte arrays of UTF-8 encoded unicode characters. (These
can be used as non-indexable arrays of bytes.)

Examples:
```
“Hello world”
  “eof”
```

### Tuples

Tuples have an arity indicating the number of elements in the
tuple. Each element in the tuple has a type. The 0-tuple is also
sometimes called unit.

Examples:
```
 {}                Type: unit
 {1, 2}            Type: Tuple(Integer, Integer)
 {“foo”, 42, true} Type: Tuple(String, Integer, Boolean)
```

### Lists

Lists are monomorphic series of values. Hence lists have a type. There
is also an empty list (or Nil).

Examples:
```
[]                   Type: Nil
[1, 2, 3]            Type: List(Integer)
[true, true, false]  Type: List(Boolean)
```

### Maps

Maps are monomorphic key value stores, that is the key is always the
same type for all elements in a map and the value has the same type in
all mappings in a map.  The key can have any type except a map.  The
value can have any type including a map.

Examples:
```
#{ (1, “foo”), (2, “bar”) }
```
Type: Map(Integer, String)
```
#{ (“Fruit prices”, #{ (“banana”, 42), (“apple”, 12)}),
   (“Stock prices”, #{(“Apple”, 142), (“Orange”, 112)}) }

> ```

Reviewer 3: Add the type for the second one as well?

Reviewer 4: `Type: Map(String, Map(String, Integer))`

# FATE
The Fast æternity Transaction Engine

## Design

> The high level machine (or the Fast æternity Transaction Engine) has

Reviewer 4: ```suggestion
The high level machine (or the Fast Æternity Transaction Engine) has
```

Author: According to our "PR department" we write aeternity in lower case.

Author: Not sure how to get the capital A in though ;)

Reviewer 4: Well, logic and PR, don't try to mix those :-)

aeternity transactions as its basic operations and it operates
directly on the state tree of the aeternity chain. This is a new
paradigm in blockchain virtual machine specifications which makes it
possible to create type safe and efficient implementations of the
machine.

Every operation is typed and all values are typed (either by being
stored in a typed memory or by a tag). Any type violation results in
an exception and reverts all state changes. In version 1.0 there will
be no catch instruction.

In addition to normal machine instructions, such as ADD, the machine

> also has support for constructing all the transactions available on

Reviewer 4: ```suggestion
also has support for constructing most of the transactions available on
```

the æternity chain from native low level chain transaction
instructions.

The instruction memory is divided into functions and basic
blocks. Only basic blocks can be indexed and used as jump
destinations.

There are instructions to operate on the chain state tree in a safe and formalized way.

FATE is “functional” in the sense that “updates” of data structures,
such as tuples, lists or maps does not change the old values of the
structure, instead a new version is created. Unless specific
operations to write to the contract store are used.

FATE does have the ability to write the value of an operation back to the
same register or stack position as one of the arguments, in effect updating
the memory. Still, any other references to the structure before the operation
will have the same structure as before the operation.

### Objectives

#### Type safety

FATE solves a fundamental problem programmers run into when coding for

> Ethereum: integer overflow, weak type checking and poore data

Reviewer 4: ```suggestion
Ethereum: integer overflow, weak type checking and poor data
```

* Memory: Several components as defined below
* CF, current function: Hash of current function
* CBB, current basic block: Index of current basic block.
* EC, Execution continuation: A list of instructions left in the current basic block.

### Memory
The machine memory is divided into:
* The chain top (block height, etc)
* Event stream
* The account state tree
* The contract store
* The execution/code memory
* Execution stack (push/pop on call/return)
* Accumulator stack (push/pop on instruction level)
* Local storage/stack/environment/context (push/pop on let...in...end)

#### The execution memory

The code to execute is stored in the execution memory. The execution
memory is a three level map. The key on the top level is a contract
address. The second level is a map from function hashes to a map of
basic blocks. The keys in the map of basic blocks are indices to basic
blocks (a 0 based, dense enumeration). The values are lists of
instructions.

When a contract is called the code of the contract is read from the
state tree into the execution memory. At this point the machine
implementation can deserialize the serialized version of FATE code
into instructions and basic blocks.

> The serialization format for FATE code is described in Appendix 1:

Reviewer 4: maybe a link?

The local storage is a key value store.
The key is a an integer. The value can be of any FATE type.
The local store is local to a function call.

#### The accumulator stack

The accumulator stack is a stack of typed values. The top of the stack
can be used as an argument to an operation. Values can be pushed to
the stack and popped from the stack.

#### The execution stack

The execution stack contain return addresses as a tuple in the form of
{contract address, function hash, and basic block index}.

#### The event stream

A contract can emit events similar to the EVM.

## Types

The machine supports the following types:
* Integer (signed arbitrary precision integers)
* Boolean (true | false)
* Strings (arbitrary size byte arrays)
* Bytes (fixed size byte arrays)
* Bits (An arbitrary size bitmap)
* Address (A pointer into the state tree)
* Contract Address (A pointer to a contract)
* Oracle (A pointer to an oracle)
* Oracle Query (A pointer to an oracle query)
* Channel (A pointer to a channel)
* Tuples (a typed series of values)
* Lists (a list of values of a specific type)
* Maps (a key value store)
* Store Map (a key value store mapped to the contract state)

> * Variant Types (a set of taged and typed values)

Reviewer 4: ```suggestion
* Variant Types (a set of tagged and typed values)
```

| 0x78 |            'ORACLE_CHECK' |   false |   false |     true |      100 |
| 0x79 |      'ORACLE_CHECK_QUERY' |   false |   false |     true |      100 |
| 0x7a |               'IS_ORACLE' |   false |   false |     true |      100 |
| 0x7b |             'IS_CONTRACT' |   false |   false |     true |      100 |
| 0x7c |              'IS_PAYABLE' |   false |   false |     true |      100 |
| 0x7d |                 'CREATOR' |   false |    true |     true |       10 |
| 0x7e |      'ECVERIFY_SECP256K1' |   false |    true |     true |     1300 |
| 0x7f |     'ECRECOVER_SECP256K1' |   false |    true |     true |     1300 |
| 0xfa |              'DEACTIVATE' |   false |    true |     true |       10 |
| 0xfb |                   'ABORT' |    true |    true |     true |       10 |
| 0xfc |                    'EXIT' |    true |    true |     true |       10 |
| 0xfd |                     'NOP' |   false |    true |     true |        1 |

## Gas
Each instruction uses the base gas as described in the table above.
In addition the instructions uses gas in relation to the number of memory
cells needed to store the produced value of the instruction.

> In addition to the instruction base cost some instructions also

Reviewer 4: This paragraph and the previous overlap?

cost gas in relation to the memory they use. The base cost for
memory is one gas per "cell" used. A cell is the memory word of
an underlying machine which is defined to be a 64-bit word.

Each use of a a cell is counted and for each 1024 cells used
the price is increased by 1. If a contract uses 1024 cells it
will cost 1024 gas, if it uses 1025 cells it will cost 1026 gas.
Using 2049 cells costs 3072 gas and so on.

When calculating the cell cost the total number of cells
used after the instruction is used to determine the price for all
new cells used by the instruction. So if you have used 1020 cells
and then one instruction uses 5 new cells, then the gas cost
for that instruction is 10 and not 6.

> Each use of a stack slot (a psuh) costs gas in the same way and

Reviewer 4: ```suggestion
Each use of a stack slot (a push) costs gas in the same way and
```

a stack slot does not cost gas and does not decrease the gas cost
either but it reduced the count of number of cells in use.

The instructions uses cells in the following way.

### ADD, SUB, MUL, DIV, MOD
After the operation is done the size of the absolute value of the
result is calculated, and one cell is used per 64 bits that has a 1 set.
Integers -18446744073709551615 to 18446744073709551615 uses 1 cell,
and integers -340282366920938463463374607431768211456 to -18446744073709551616
and 18446744073709551616 to 340282366920938463463374607431768211456
uses 2 cells, and so on.

### POW
The result of the pow instruction is computed recursively
and for each step in the recursion the words used are
calculated and the operation is aborted if all gas is used up.
Given the function words which calculates the number of cells in
an integer as described above (1 per 64 bits used).
The number of cells are calculated as follows.

```
pow(a, b)    := pow(a, b, 1)

pow(a, b, r) := r  when b = 0
pow(a, b, r) := cells += words(a) * 2, pow(a*a, b/2, r) when b rem 2 = 0
pow(a, b, r) := cells += words(r) + words(a) * 3, pow(a*a, b/2, r*a)
```

### TUPLE

> For creating a tuple you have already paied gas for the elements when

Reviewer 4: ```suggestion
For creating a tuple you have already payed gas for the elements when
```

you pushed them on the stack with other operations, then the new tuple

> increases the cell count by the size of the tuple plus 2.

Reviewer 4: ```suggestion
increase the cell count by the size of the tuple plus 2.
```

The number of cells used is the tuple size + 2.

### MAP_EMPTY
The cell cost is 2.

### MAP_UPDATE
The cell cost is 2.

### MAP_FROM_LIST
The cell cost is 2 + length of the list.

### MAP_TO_LIST
The cell cost is 2 * length of the list.

### CONS
The cell cost is 2.

### APPEND
The cell cost is 2 * length of the first list.

### STR_JOIN
The cell cost is 1 cell per 8 bytes in the joined string + 1 cell.

### INT_TO_STR
The cell cost is 1 cell per 8 bytes in the produced string + 1 cell.

### ADDR_TO_STR
The cell cost is 1 cell per 8 bytes in the produced string + 1 cell.

### STR_REVERSE
The cell cost is 1 cell per 8 bytes in the produced string + 1 cell.

### INT_TO_ADDR
The cell cost is 1 cell per 8 bytes in the produced address + 1 cell.

### BITS_NONE, BITS_NONEA, BITS_ALL, BITS_ALLA
The cell cost is 1 cell.

### BITS_ALL_N
The sell cost is 1 for every 64 bits used.

### BITS_SET, BITS_CLEAR
The sell cost is 1 for every 64 bits used by the resulting bit field.

### BYTES_CONCAT
The cell cost is 1 cell per 8 bytes in the produced byte array.

### VARIANT
The cell cost is two times the length of the list of arities plus one for
each element in the variant plus four.

> 

Reviewer 4: BYTES_TO_STR and BYTES_SPLIT are missing.

FATE does have the ability to write the value of an operation back to the
same register or stack position as one of the arguments, in effect updating
the memory. Still, any other references to the structure before the operation
will have the same structure as before the operation.

### Objectives

#### Type safety

FATE solves a fundamental problem programmers run into when coding for
Ethereum: integer overflow, weak type checking and poor data
flow. FATE checks all aritmetic operations to keep the right meanining
of it. Also you can't implicitly cast types (eg integers to booleans).

In EVM contracts are not typed. When calling a function, EVM will try
to find which function you wanted to use by looking at the data that
you send into the code. In FATE, you have actual functions and the
functions have types - function call has to match the function type.

FATE ultimately makes a safer coding platform for smart contracts.

#### Rich data types

FATE has more built in data types: like maps, lists.

#### Faster transactions, smaller code size

Having a higher level instructions makes the code deployed smaller and
it reduces the blockchain size. Smaller code and smaller hashes also

> keeps a blockchain network from clogging up.

Reviewer 5: Because of this, transactions also are cheaper. Is it worth mentioning it? I personally think that this is more important than the clogging statement.

Examples:
```
"Hello world"
  "eof"
```

### Tuples

Tuples have an arity indicating the number of elements in the
tuple. Each element in the tuple has a type. The 0-tuple is also
sometimes called unit.
The maximum number of elements in the tuple is 255 (included).

Examples:
```
 {}                Type: unit
 {1, 2}            Type: Tuple(Integer, Integer)
 {"foo", 42, true} Type: Tuple(String, Integer, Boolean)
```

### Lists

Lists are monomorphic series of values. Hence lists have a type. There
is also an empty list (or Nil).

Examples:
```
[]                   Type: Nil
[1, 2, 3]            Type: List(Integer)
[true, true, false]  Type: List(Boolean)
```

### Maps

Maps are monomorphic key value stores, that is the key is always the
same type for all elements in a map and the value has the same type in
all mappings in a map.  The key can have any type except a map.  The
value can have any type including a map.

Examples:
```
#{ (1, "foo"), (2, "bar") }
```
Type: Map(Integer, String)

```
#{ ("Fruit prices", #{ ("banana", 42), ("apple", 12)}),
   ("Stock prices", #{("Apple", 142), ("Orange", 112)}) }
```
Type: Map(String, Map(String, Integer))

### Variant Types

> The variant type is a type consisting of a list oif sizes and a tag (index)

Reviewer 6: ```suggestion
The variant type is a type consisting of a list of sizes and a tag (index)
```
